api/proto/teleport/userintegrationtasks/v1/user_integration_tasks.proto

```diff
@@ -0,0 +1,84 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package teleport.userintegrationtasks.v1;
+
+import "google/protobuf/timestamp.proto";
+import "teleport/header/v1/metadata.proto";
+
+option go_package = "github.com/gravitational/teleport/api/gen/proto/go/teleport/userintegrationtasks/v1;userintegrationtasksv1";
+
+// UserIntegrationTask is a resource that represents an action to be completed by the user.
+// UserIntegrationTasks are a unit of work for users to act upon issues related to integrations.
+// As an example, when auto-enrolling EC2 instances using the Discovery Service
+// a UserIntegrationTask is created to let the user know that something failed on a set of instances.
+// The user can then mark the task as resolved after following the recommendation/fixing steps.
+message UserIntegrationTask {
+  // The kind of resource represented.
+  string kind = 1;
+  // Mandatory field for all resources. Not populated for this resource type.
+  string sub_kind = 2;
+  // The version of the resource being represented.
+  string version = 3;
+  // Common metadata that all resources share.
+  teleport.header.v1.Metadata metadata = 4;
+  // The configured properties of UserIntegrationTask.
+  UserIntegrationTaskSpec spec = 5;
+}
+
+// UserIntegrationTaskSpec contains the properties of the UserIntegrationTask.
+message UserIntegrationTaskSpec {
+  // Integration is the integration name that originated this task.
+  string integration = 1;
+  // TaskType indicates the type of task.
+  // Examples: discover-ec2, discover-rds, discover-eks
+  string task_type = 2;
+  // IssueType is an identifier for the type of issue that happened.
+  // Example for discover-ec2: SSM_AGENT_NOT_AVAILABLE
+  string issue_type = 3;
+  // DiscoverEC2 contains the AWS EC2 instances that failed to auto enroll into the cluster.
+  // Present when TaskType is discover-ec2.
+  DiscoverEC2 discover_ec2 = 4;
+}
+
+// DiscoverEC2 contains the instances that failed to auto-enroll into the cluster.
+message DiscoverEC2 {
+  // Instances maps an instance id to the result of enrolling that instance into teleport.
+  map<string, DiscoverEC2Instance> instances = 1;
+}
+
+// DiscoverEC2Instance contains the result of enrolling an AWS EC2 Instance.
+message DiscoverEC2Instance {
+  // State indicates the current issue state.
+  // When the issue is created, it starts with OPEN.
+  // Users can mark this instance issue as RESOLVED.
+  // If the issue happens again (eg, new discover iteration faces the same error), it will move to OPEN again.
+  string state = 1;
```

Comment:
Thinking about this, I don't know if having a task state as a part of instance makes sense. Logically, a task itself should have a state.

I propose we move state to the top level of task spec. If there are multiple instances that are failing to enroll with the same error, we can just record a list of them in the task and when it's retried, either update the task with the new list of servers or just clear it out and create a new one.

Reply:
From the user's perspective, I think it makes sense to have a state per instance.
Let's say you have 10 instances that don't have the SSM Agent running.
The way I would fix this would be to:
- click on an instance which opens a new tab in Amazon EC2 dashboard
- connect into the instance and fix it
- mark this instance as resolved in Teleport
- go back to first step

We can add a status/state to the entire task, but I'm not seeing an actual use case for it, at least for now.
We can show it to the user as the result of checking individual instances.
Eg, if all the instance's states are RESOLVED, then the Task is RESOLVED as well.
Otherwise, it's OPEN.
We can add helper buttons to the UI to ensure users can click on a button to Resolve all instances.

Reply:
After discussing this with Product, we decided to move the state/status to the task.